### PR TITLE
Resolved : JS files located outside the web/js directory

### PR DESCRIPTION
**Pull request by hitesh-wagento**
### Description
Some JS files are direct children of `web` rather than `web/js`. This does not follow instructions from the [dev docs](https://devdocs.magento.com/guides/v2.2/javascript-dev-guide/javascript/custom_js.html) thus is confusing.

### Fixed Issues (if relevant)

1. <https://github.com/magento/magento2/issues/16302>: JS files located outside the web/js directory

### Expected result
1. JS files to be inside `web/js` as per the dev docs.

### Actual result
1. JS files are direct children of `web`, this is inconsistent with the dev docs.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @hitesh-wagento. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation

**Comment by VladimirZaets** [comment]
Hi @hitesh-wagento,  due to [Magento backward-compatible guide](http://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/) we can't move or rename files. 
These changes are valid for 2.3 version, please create new one PR based on 2.3-develop branch.
Thanks

**Comment by ihor-sviziev** [comment]
Hi @hitesh-wagento,
As we discussed in https://github.com/magento/magento2/pull/16708#issuecomment-405902812 - let's make this PR backward compatible and I will process it first.

**Comment by ihor-sviziev** [comment]
Hi @hitesh-wagento,
We are having discussion about this specific case with @VladimirZaets, so please wait once we will finalize out general vision how it should be done. I will update you after discussion.

**Comment by hitesh-wagento** [comment]
Hi @ihor-sviziev , @VladimirZaets 

I have completed changes as per your suggestion please review and let me know if any changes required.

Thanks

**Comment by VladimirZaets** [comment]
Hi, @hitesh-wagento thanks for fixes. Please resolve the branch conflicts. As I see you apply the fix that was provided by @ihor-sviziev 
```
 map: {
         '*': {
            transparent: 'Magento_Payment/transparent'
            transparent: 'Magento_Payment/js/transparent',
            'Magento_Payment/transparent': 'Magento_Payment/js/transparent'
         }
     }
```

you should to do it for each cases. 
As example, currently you have 
```
map: {
         '*': {
            captcha: 'Magento_Captcha/captcha'
            captcha: 'Magento_Captcha/js/captcha'
         }
     }
```

also you should add  `'Magento_Captcha/captcha': 'Magento_Captcha/js/captcha'

**Comment by hitesh-wagento** [comment]
Hi @ihor-sviziev

As per our Slack discussion I have completed changes please review and let me know if any changes required.

Thanks

File: app/code/Magento/SendFriend/view/frontend/templates/send.phtml [hunk 1]
```diff
@@ -123,7 +123,7 @@
 <script type="text/x-magento-init">
     {
         "a[role='back']": {
-            "Magento_SendFriend/back-event": {}
+            "Magento_SendFriend/js/back-event": {}
```
Comment (ihor-sviziev) [inline app/code/Magento/SendFriend/view/frontend/templates/send.phtml:126]: We need to create mapping between old and new path in requirejs-config.js file in Magento_SendFriend module

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/SendFriend/view/frontend/templates/send.phtml (hunk 1 above)
Comment (hitesh-wagento) [reply app/code/Magento/SendFriend/view/frontend/templates/send.phtml:126]: Applied changes as per our slack conversion.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-2415](https://jira.corp.magento.com/browse/ENGCOM-2415) has been created to process this Pull Request

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-2415](https://jira.corp.magento.com/browse/ENGCOM-2415) has been created to process this Pull Request

**Comment by hitesh-wagento** [comment]
Hi @ihor-sviziev , @VladimirZaets 

Any update in this PR ?

Thanks

**Comment by ihor-sviziev** [comment]
Hi @hitesh-wagento,
This PR didn't passed testing yet. That's why it didn't merged yet.

**Comment by magento-engcom-team** [comment]
Hi @hitesh-wagento. Thank you for your contribution. 
We will aim to release these changes as part of 2.2.7. 
Please check the release notes for final confirmation.

Please, consider to port this solution to 2.3 release line.
You may use [Porting tool](https://github.com/magento/magento2/wiki/Simple-way-to-port-your-solutions-automatically) to port commits automatically.

**Comment by hitesh-wagento** [comment]
Hi @ihor-sviziev

Do I need to create ports for this PR ? 
Because I have already created PR for this : https://github.com/magento/magento2/pull/16708

Thanks

**Comment by ihor-sviziev** [comment]
Hi @hitesh-wagento,
No, you should not

**Comment by keharper** [comment]
@hitesh-wagento You linked to devdocs at the top of this PR. It looks like you changed the code to meet the standards set in the documentation, but I want to verify whether you made any documentable changes.

**Comment by hitesh-wagento** [comment]
> @hitesh-wagento You linked to devdocs at the top of this PR. It looks like you changed the code to meet the standards set in the documentation, but I want to verify whether you made any documentable changes.

Hi @keharper 

I have changed  as per suggested by ihor-sviziev

**Comment by rbayet** [comment]
Hi @ihor-sviziev, @hitesh-wagento,

I'm not sure it's the best of places to address that, but I didn't want to re-open https://github.com/magento/magento2/issues/16302 by adding a comment to it.

My issue is it seems the changes seem to be BC under some conditions.

For instance, if in a 2.2.5 :
- I manually relocate Magento_Search/form-mini to Magento_Search/js/form-mini
- I apply manually the changes in Magento_Search/requirejs-config.js  
```
var config = {
    map: {
        '*': {
            quickSearch: 'Magento_Search/js/form-mini',
            'Magento_Search/form-mini': 'Magento_Search/js/form-mini'
        }
    }
};
```
- I have a third party module/theme overriding the "quickSearch" requirement and referencing the native/legacy path as "mageQuickSearch"
```
var config = {
    map: {
        '*': {
            mageQuickSearch : 'Magento_Search/form-mini',
            quickSearch: 'Third_PartyModule/js/form-mini'
        }
    }
};
```
'mageQuickSearch' being there to be referenced as a define() dependency in 'Third_PartyModule/js/form-mini'.

*Problem*: The path mapping is not taken into account, either in blank or luma, in developer or production modes : a 404 error is thrown for ".../Magento_Search/form-mini.js".

*IF* I remove "mageQuickSearch" from the third party requirejs-config.js and replace it in the 'Third_PartyModule/js/form-mini' define() by the native/legacy path 'Magento_Search/form-mini', then it works as expected.

Am I missing something or does that mean there is still a risk of BC break across the board ?

Regards

**Pull request merged**